File: crates/core/src/protocol.rs
//! WebSocket wire protocol between daemon and SPA.
//! Message #1 in each direction is the version handshake — the hosted SPA
//! updates instantly, installed daemons don't, so skew is a permanent state
//! to handle, not an error to hide.

use crate::hunk::{FileDiff, HunkId};
use crate::reconcile::ReconcileReport;
use crate::review::ReviewState;
use crate::schema::{ImpactScore, Walkthrough};
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

#[cfg(feature = "ts-export")]
use ts_rs::TS;

pub const PROTOCOL_VERSION: u16 = 5;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
#[cfg_attr(feature = "ts-export", derive(TS))]
pub enum ClientMsg {
    /// Must be the first message. Token echoes the URL fragment value.
    Hello {
        protocol: u16,
        token: String,
    },
    MarkViewed {
        hunk: HunkId,
    },
    AddFlag {
        hunk: HunkId,
        /// Index into the hunk's raw lines this comment anchors to, GitHub
        /// per-line style. None = a hunk-level comment. Identity is still the
        /// hunk (invariant 2); this is only a position within it.
        #[serde(default)]
        line: Option<u32>,
        comment: String,
    },
    CloseFlag {
        hunk: HunkId,
        #[serde(default)]
        line: Option<u32>,
    },
    /// Anchored prompt dispatch — always attached to hunks, never free-form.
    RequestChange {
        hunks: Vec<HunkId>,
        /// Review thread anchor being dispatched. Claims return only to this
        /// thread, never every open flag sharing the hunk.
        #[serde(default)]
        line: Option<u32>,
        instruction: String,
        runner: String,
    },
    ApplyUpdate {
        #[cfg_attr(feature = "ts-export", ts(type = "number"))]
        to_revision: u64,
    },
    Regenerate,
    ExportReview,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
#[cfg_attr(feature = "ts-export", derive(TS))]
pub enum ServerMsg {
    HelloAck {
        protocol: u16,
        daemon_version: String,
        /// Human label for the walkthrough organizer, e.g. "claude (your
        /// login)" or "none (deterministic fallback)".
        llm: String,
    },
    /// Full state on connect / after ApplyUpdate. Browser holds no durable state.
    Snapshot {
        walkthrough: Walkthrough,
        files: Vec<FileDiff>,
        scores: BTreeMap<HunkId, ImpactScore>,
        review: ReviewState,
    },
    /// Background reconciliation result — UI shows the apply banner,
    /// never reflows on its own.
    UpdateAvailable {
        #[cfg_attr(feature = "ts-export", ts(type = "number"))]
        revision: u64,
        report: ReconcileReport,
    },
    GenerationProgress {
        message: String,
    },
    /// Review state changed (comment added, flag closed, hunk viewed) without
    /// the diff moving. Pushed so every tab reflects it immediately — the
    /// browser holds no durable state, so a mutation the daemon doesn't echo
    /// is invisible until reconnect. Lighter than a full Snapshot.
    ReviewUpdated {
        review: ReviewState,
    },
    DispatchStatus {
        job_id: String,
        status: JobStatus,
        /// Human-readable note: the agent's change summary on Done, the
        /// revert reason on failure, the out-of-scope files on
        /// ScopeViolation. None while merely Running.
        detail: Option<String>,
    },
    ReviewExport {
        markdown: String,
    },
    Error {
        code: ErrorCode,
        message: String,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
#[cfg_attr(feature = "ts-export", derive(TS))]
pub enum JobStatus {
    Running,
    Done,
    Failed,
    TimedOutReverted,
    /// Agent touched files outside the anchored set — surfaced, not hidden.
    ScopeViolation,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
#[cfg_attr(feature = "ts-export", derive(TS))]
pub enum ErrorCode {
    BadToken,
    ProtocolMismatch,
    BusyWriterLock,
    Internal,
}

File: web/src/libs/protocol.ts
// Wire protocol — mirrors crates/core/src/protocol.rs (PROTOCOL_VERSION 1).
// TODO(M1): replace this hand-written mirror with ts-rs codegen from the
// Rust structs (`cargo test --features ts-export` emits bindings). Until
// then, this file is the single place to keep in sync.

export const PROTOCOL_VERSION = 4;

export type Impact = "low" | "medium" | "high" | "highest";

export interface ImpactScore {
  impact: Impact;
  points: number;
  reasons: string[];
}

export interface Hunk {
  id: string;
  path: string;
  new_start: number;
  old_start: number;
  added: number;
  removed: number;
  lines: string[];
}

export interface FileDiff {
  path: string;
  old_path: string | null;
  status: "added" | "modified" | "deleted" | "renamed";
  hunks: Hunk[];
}

export interface Step {
  id: string;
  title: string;
  framing: string;
  hunks: string[];
}

export interface Scope {
  id: string;
  title: string;
  steps: Step[];
}

export interface Walkthrough {
  schema_version: number;
  revision: number;
  tree_state: string;
  /** Reading-order description ("review focus"). Null in fallback mode. */
  focus: string | null;
  scopes: Scope[];
  degraded: boolean;
}

export type HunkStatus = "unviewed" | "viewed" | "changed_since_viewed";

/** No "verdict" kind exists on purpose — the machine never judges the code. */
export type FlagEntryKind = "human_comment" | "agent_claim" | "dispatch_note";

export interface FlagEntry {
  kind: FlagEntryKind;
  body: string;
  revision: number;
}

export interface Flag {
  hunk: string;
  /** Index into hunk.lines this thread anchors to (per-line comment), or
   *  null for a hunk-level comment. Render offset only — identity is the hunk. */
  line: number | null;
  /** Oldest-first: the human's comment, agent change-claims, dispatch notes. */
  thread: FlagEntry[];
  open: boolean;
  addressed_claim: boolean;
}

export interface ReviewState {
  status: Record<string, HunkStatus>;
  flags: Flag[];
  tombstones: Flag[];
}

export interface Lineage {
  from: string;
  to: string;
}

export interface ReconcileReport {
  carried: string[];
  changed: Lineage[];
  added: string[];
  removed: string[];
}

export type ClientMsg =
  | { type: "hello"; protocol: number; token: string }
  | { type: "mark_viewed"; hunk: string }
  | { type: "add_flag"; hunk: string; line: number | null; comment: string }
  | { type: "close_flag"; hunk: string; line: number | null }
  | {
      type: "request_change";
      hunks: string[];
      line: number | null;
      instruction: string;
      runner: string;
    }
  | { type: "apply_update"; to_revision: number }
  | { type: "regenerate" }
  | { type: "export_review" };

export type ServerMsg =
  | { type: "hello_ack"; protocol: number; daemon_version: string; llm: string }
  | {
      type: "snapshot";
      walkthrough: Walkthrough;
      files: FileDiff[];
      scores: Record<string, ImpactScore>;
      review: ReviewState;
    }
  | { type: "update_available"; revision: number; report: ReconcileReport }
  | { type: "generation_progress"; message: string }
  | { type: "review_updated"; review: ReviewState }
  | {
      type: "dispatch_status";
      job_id: string;
      status: "running" | "done" | "failed" | "timed_out_reverted" | "scope_violation";
      detail: string | null;
    }
  | { type: "review_export"; markdown: string }
  | { type: "error"; code: string; message: string };
